src/Test/QuickCheck/Test.hs

```diff
@@ -228,9 +228,16 @@ withState a test = (if chatty a then withStdioTerminal else withNullTerminal) $
           -- 0, 1, 2, ..., 99, 0, 1, 2, ..., 99, 0, 2, 4, ..., 98.
           | n `roundTo` maxSize a + maxSize a <= maxSuccess a ||
             n >= maxSuccess a ||
-            maxSuccess a `mod` maxSize a == 0 = (n `mod` maxSize a + d `div` 10) `min` maxSize a
+            maxSuccess a `mod` maxSize a == 0 = (n `mod` maxSize a + d `div` dDenom) `min` maxSize a
           | otherwise =
-            ((n `mod` maxSize a) * maxSize a `div` (maxSuccess a `mod` maxSize a) + d `div` 10) `min` maxSize a
+            ((n `mod` maxSize a) * maxSize a `div` (maxSuccess a `mod` maxSize a) + d `div` dDenom) `min` maxSize a
+        -- The inverse of the rate at which we increase size as a function of discarded tests
+        -- if the discard ratio is high we can afford this to be slow, but if the discard ratio
+        -- is low we risk bowing out too early
+        dDenom
+          | maxDiscardRatio a > 0 = (maxSuccess a * maxDiscardRatio a `div` 3) `clamp` (1, 10)
+          | otherwise = 1 -- Doesn't matter because there will be no discards allowed
+        clamp x (l, h) = min l (max x h)
```

Comment:
```suggestion
        clamp x (l, h) = max l (min x h)
```

Reply:
🤦 50% probability of getting it 100% wrong 50% of the time...